modules/cluster/variables.tf

```
@@ -189,6 +189,30 @@ variable "critical_addons_node_group_bottlerocket_admin_container_source" {
   description = "URI of a custom admin container image"
 }
 
+variable "critical_addons_vpc-cni_configuration_values" {
+  type        = string
+  default     = null
+  description = "Configuration values passed to the vpc-cni EKS addon."
+}
+
+variable "critical_addons_kube-proxy_configuration_values" {
+  type        = string
+  default     = null
+  description = "Configuration values passed to the kube-proxy EKS addon."
+}
+
+variable "critical_addons_coredns_configuration_values" {
+  type        = string
+  default     = null
```

We could use structural typing here to do some kind of validation on the inputs... unless this is already handled internally by the provider???

```suggestion
  type = object({ computeType = string, corefile=string, nodeSelector=object(string), replicaCount=number, resources=object(string)})
  default  = null
```

---

Different versions of an EKS addon have different schema for their configuration values - so the provider simply transparently passes the json-encoded string, and leaves mapping the addon version <-> configuration value schema to the user.

Because we hard-code the addon versions, it _is_ technically possible for us to provide a well typed interface with the configuration values. I will do this.

I should note a drawback that this adds more steps to our upgrade process for this module. We will need through each addon version we've updated, fetch the latest configuration values schema, and map that into the variables.

We should probably do this activity anyway however, to notify users of breaking changes in the configuration value schema.


> :memo: Note to self: command to check configuration value schema is
> ```
> $ envchain aws aws --profile=cookpad-global-1 eks describe-addon-configuration --addon-name aws-ebs-csi-driver --addon-version v1.10.0-eksbuild.1 | jq '.configurationSchema' -r | jq
> ```

---

If we do this I think we could look at adding a unit test to check our well typed interface against the jsonschema for the addon

---

I think we can use https://pkg.go.dev/github.com/hashicorp/terraform-config-inspect/tfconfig#LoadModule within a unit test to load [the cluster module and its Variables](https://pkg.go.dev/github.com/hashicorp/terraform-config-inspect/tfconfig#Module)

---

I chose a different direction - see https://github.com/cookpad/terraform-aws-eks/pull/331#issuecomment-1428155603